qualtran/bloqs/controlled_state_preparation.py

```diff
@@ -0,0 +1,483 @@
+need of a $R_x(\pm \pi/2)$ gate before and after encoding the amplitudes of each qubit.
+
+In order to perform the rotations as efficiently as possible, the angles are loaded into a register
+(rot\_reg) which is added into a phase gradient. Then phase kickback causes an overall offset of
+$e^{i2\pi x/2^b}$, where $x$ is the angle value loaded and $b$ the size of the rot\_reg. Below is an
+example for rot\_reg\_size=2.
+
+First there is the rot\_reg register with the value to be rotated (3 in this case) and the phase
+gradient
+
+$$
+|3\rangle(e^{2\pi i 0/4}|0\rangle + e^{2\pi i 1/4}|1\rangle +
+          e^{2\pi i 2/4}|2\rangle + e^{2\pi i 3/4}|3\rangle).
+$$
+
+Then the rot\_reg $|3\rangle$ register is added to the phase gradient and store the result in the
+phase gradient register
+
+$$
+|3\rangle(e^{2\pi i 0/4}|3\rangle + e^{2\pi i 1/4}|0\rangle +
+          e^{2\pi i 2/4}|1\rangle + e^{2\pi i 3/4}|2\rangle),
+$$
+
+but this is equivalent to the original state with a phase offset of $e^{2\pi i 1/4}$.
+
+
+References:
+    [Trading T-gates for dirty qubits in state preparation and unitary synthesis]
+    (https://arxiv.org/abs/1812.00954).
+        Low, Kliuchnikov, Schaeffer. 2018.
+
+"""
+
+from typing import Dict, Tuple
+
+import attrs
+import numpy as np
+from numpy.typing import ArrayLike
+
+from qualtran import (
+    Bloq,
+    bloq_example,
+    BloqBuilder,
+    BloqDocSpec,
+    SelectionRegister,
+    Signature,
+    SoquetT,
+)
+from qualtran.bloqs.arithmetic import Add
+from qualtran.bloqs.basic_gates import OneEffect, OneState, ZeroEffect, ZeroState
+from qualtran.bloqs.basic_gates.rotation import Rx
+from qualtran.bloqs.qrom import QROM
+from qualtran.bloqs.select_and_prepare import PrepareOracle
+
+
+@attrs.frozen
+class ControlledStatePreparationUsingRotations(PrepareOracle):
+    r"""Controlled state preparation without entangled residual using Ry and Rz rotations from [1].
+
+    Given a quantum state of which the list of coefficients $c_i$ is known
+    $$
+        |\psi \rangle = \sum_{i=0}^{N-1}c_{i}|i\rangle
+    $$
+    this gate prepares $|\psi\rangle$ from $|0\rangle$ conditioned by a control qubit
+    $$
+        U((|0\rangle + |1\rangle)|0\rangle) = |0\rangle |0\rangle + |1\rangle |\psi\rangle.
+    $$
+
+    Args:
+        state_bitsizes: number of qubits of the state.
+        rot_reg_bitsizes: size of the register that is used to store the rotation angles. Bigger values
+            increase the accuracy of the results.
+        state: tuple of length 2^state_bitsizes that contains the complex coefficients of the state.
+
+    References:
+        [Trading T-gates for dirty qubits in state preparation and unitary synthesis]
+        (https://arxiv.org/abs/1812.00954).
+            Low, Kliuchnikov, Schaeffer. 2018.
+    """
+
+    state_bitsizes: int
+    rot_reg_bitsizes: int
+    state: Tuple
+    uncompute: bool = False
```

Comment:
Do we imagine a scenario where `len(state) != 2^{state_bitsize}` ? If not, then we can only ask the user to specify the `state: Tuple[complex, ...]` and we can infer the required bitsize for the register using `(len(state) - 1).bit_length()`

Another nit would be to rename the attribtutes as follows:
```suggestion
    state_bitsize: int
    state_coefficients: Tuple[complex, ...]
    phase_bitsize: int
    uncompute: bool = False
```

Reply:
Silly question: as far as I've seen, no bloq has a constructor. Where should I place code that must be called when the object is initialized? I also want to run some checks on the parameters to ensure the user utilizes the bloq correctly, but so far I have no idea where to place it.

Reply:
> as far as I've seen, no bloq has a constructor

That's because most Bloqs are annotated with `@attr.frozen` which automatically generates a constructor for the Bloq. And the auto generated constructor is super simple - just expects and assigns values to `self.<attribute_name>` where `<attribute_name>` corresponds to the list of attributes we define on top of the bloq. In this case, `state_bitsize` etc. See https://www.attrs.org/en/stable/examples.html for more details. 

The philosophy is that we almost always want to keep the constructors super simple (no fancy logic) and using the auto generated constructors enforces this along with other nice things like the generated class is hashable etc.

> I also want to run some checks on the parameters to ensure the user utilizes the bloq correctly

You can use validators for this purpose. See https://www.attrs.org/en/stable/examples.html#val and https://github.com/quantumlib/Qualtran/blob/28c0c1d3651120f3bb2f25dd40346ae4e94585e0/qualtran/bloqs/qrom.py#L85 for examples of how it's done.